packages/style-engine/class-wp-style-engine.php

```diff
@@ -186,71 +203,136 @@ protected static function get_classnames( $style_value, $style_definition ) {
 	 *     'classnames' => (string) Classnames separated by a space.
 	 * );
 	 */
-	public function generate( $block_styles ) {
+	public function generate( $block_styles, $options ) {
 		if ( empty( $block_styles ) || ! is_array( $block_styles ) ) {
 			return null;
 		}
 
-		$css_rules     = array();
-		$classnames    = array();
-		$styles_output = array();
+		$css_rules                               = array();
+		$classnames                              = array();
+		$should_generate_classnames_from_presets = isset( $options['classnames'] ) && true === $options['classnames'];
+		$should_enqueue_block_support_styles     = isset( $options['enqueue_block_support_styles'] ) && true === $options['enqueue_block_support_styles'];
 
 		// Collect CSS and classnames.
 		foreach ( self::BLOCK_STYLE_DEFINITIONS_METADATA as $definition_group ) {
+			if ( ! $definition_group ) {
+				continue;
+			}
+
 			foreach ( $definition_group as $style_definition ) {
 				$style_value = _wp_array_get( $block_styles, $style_definition['path'], null );
 
 				if ( empty( $style_value ) ) {
 					continue;
 				}
 
-				$classnames = array_merge( $classnames, static::get_classnames( $style_value, $style_definition ) );
-				$css_rules  = array_merge( $css_rules, static::get_css( $style_value, $style_definition ) );
+				// Generate classnames from var:? values.
+				if ( $should_generate_classnames_from_presets ) {
+					$classnames = array_merge( $classnames, static::get_classnames( $style_value, $style_definition ) );
+				}
+
+				$css_rules = array_merge( $css_rules, static::get_css( $style_value, $style_definition, ! $should_generate_classnames_from_presets ) );
 			}
 		}
 
 		// Build CSS rules output.
-		$css_output = '';
+		$selector      = isset( $options['selector'] ) ? $options['selector'] : null;
+		$css           = array();
+		$styles_output = array();
+
 		if ( ! empty( $css_rules ) ) {
 			// Generate inline style rules.
-			// In the future there might be a flag in the option to output
-			// inline CSS rules (for HTML style attributes) vs selectors + rules for style tags.
 			foreach ( $css_rules as $rule => $value ) {
 				$filtered_css = esc_html( safecss_filter_attr( "{$rule}: {$value}" ) );
 				if ( ! empty( $filtered_css ) ) {
-					$css_output .= $filtered_css . '; ';
+					$css[] = $filtered_css . ';';
 				}
 			}
 		}
 
-		if ( ! empty( $css_output ) ) {
-			$styles_output['css'] = trim( $css_output );
+		// Return css, if any.
+		if ( ! empty( $css ) ) {
+			if ( $selector ) {
+				$style_block          = "$selector { ";
+				$style_block         .= implode( ' ', $css );
+				$style_block         .= ' }';
+				$styles_output['css'] = $style_block;
+
+				// Enqueue block support styles where there is a selector.
+				if ( $should_enqueue_block_support_styles ) {
+					if ( isset( $this->registered_block_support_styles[ $selector ] ) ) {
```

Comment:
What do you reckon about this @andrewserong ?

I know you tried something similar over in https://github.com/WordPress/gutenberg/pull/38974

Too much too soon?

My idea was that we could leverage registration for all block styles, mainly layouts, but also consolidating elements as they're added.

Reply:
> Too much too soon?

I think this PR is probably a good time for us to start exploring it, given that this is the first block support we're dealing with refactoring that depends on rendering in a separate `style` tag.

> I know you tried something similar over

The thing I hadn't looked at yet was how will we want to call the style engine from the theme JSON class? Since enqueueing / outputting is already happening over there, I wasn't sure if we'd move that to the style engine, too, eventually, or if we'll leave that class to handle it. In which case, I guess it'd be calling the style engine without the `enqueue_block_support_styles` option, so it's probably fine?

> My idea was that we could leverage registration for all block styles, mainly layouts, but also consolidating elements as they're added.

That sounds good to me — with the slight caveat that we'd ideally be moving a bunch of layout styles to be generated via presets (as in https://github.com/WordPress/gutenberg/pull/39708). So, I imagine ideally via the style engine, we'll be mostly handling real values that need to be generated rather than common styles for output? Or were you imagining we'd deal with those here, too?

Reply:
> The thing I hadn't looked at yet was how will we want to call the style engine from the theme JSON class?

Thanks for the comment!

I'm playing (only playing) around with that over in https://github.com/WordPress/gutenberg/pull/40955

Reply:
> I'm playing (only playing) around with that over in https://github.com/WordPress/gutenberg/pull/40955

Nice one! Well, that seems to confirm that guarding the behaviour behind the `$should_enqueue_block_support_styles` option here means that the implementation in global styles output can happily co-exist with enqueuing for when individual block supports are rendered.

Just an idle thought, but in terms of naming I wonder if we can remove the `block_support` from the option name and just call it `enqueue_styles` since theoretically we could one day use the enqueuing outside of block supports (if it winds up being better to use the enqueuing for the global styles output, too, for example)?

Reply:
> if we can remove the block_support from the option name and just call it enqueue_styles since theoretically we could one day use the enqueuing outside of block supports

Terrific question. 

I agree. Ultimately we could (and probably should?) work towards rendering global styles in the same block, so the name wouldn't be appropriate. 

I did also wonder about this at the time. The reason I kept the `block_supports` in the name for now was because there appears to be a deliberate distinction between global styles and block styles in https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-5.9/script-loader.php#L63

So I ~stole~ borrowed the code from `gutenberg_enqueue_block_support_styles()`, which doesn't do half the work of `gutenberg_enqueue_global_styles_assets()` 

So I suppose it was mainly to remind me not to load anything else without checking first 😄 Plus I reasoned it was a private method and therefore "renameable".

Reply:
> there appears to be a deliberate distinction between global styles and block styles

Oh, great point (and about it mostly being contained in private method names, aside from the `$options` param). I think given that we're still in "refactor working features" phase of the style engine work, keeping the consistency with the existing behaviour there sounds like a good way to do it.

And like you say, if and when we get to consolidating with global styles, we can combine the two as needed when we get there 👍

Reply:
> and JS later?

Definitely 🙂

If we ever want to add client-side transitions, we need to know the CSS+JS requirements for each block, whether those are external assets (URLs), their own `script`/`style` tags or references to other (more global) `script`/`style` tags. And the same is true for the REST API ([the famous issue](https://github.com/WordPress/gutenberg/issues/35376)), and could be a way to ease the integration of Gutenberg on headless sites.